numpy/core/src/multiarray/descriptor.c

```diff
@@ -2797,9 +2797,55 @@ arraydescr_setstate(PyArray_Descr *self, PyObject *args)
                 }
             }
 #else
-            PyErr_Format(PyExc_ValueError,
-                "non-string names in Numpy dtype unpickling");
-            return NULL;
+            // This section is an adaption of the block above (which reads PY2 pickles from PY3)
+            // to enable reading PY3 pickles from PY2. See gh-2407
+            PyObject *tmp, *new_name, *field;
```

Comment:
I think we are still c-89 compliant with the 1.16 branch, so you cannot define variables in the middle of a block. Define `new_name` and `field` in the loop, and `tmp` at the beginning of the current block.

Reply:
Ahh, I see that was wrong and the problem was a missing closing '}'

Reply:
Yea, I had missed it originally when I moved the code to the 1.16 branch but added it to the second commit. Thanks for the help!